### ISSUE TITLE
False positive: Recommended shorthand for non-commutative multiplication

### ISSUE THREAD

**Issue by Limeth** [rust-lang-nursery/rust-clippy#2266]
Clippy reports the following warning:

```
warning: manual implementation of an assign operation
   --> src/util.rs:658:21
    |
658 |                     *vector = result * *vector;
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `*vector *= result`
...
665 | dimension!(Point2, Vector2, Matrix2, 2);
    | ---------------------------------------- in this macro invocation
    |
    = note: #[warn(assign_op_pattern)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.175/index.html#assign_op_pattern
```

where `vector` is of type `nalgebra::Vector2<F>` and `result` is of type `nalgebra::Matrix2<F>`.
Clippy suggests changing the code to `*vector *= result`, but that desugars to `*vector = *vector * result`, but since vector-matrix multiplication is not commutative, the resulting value of `vector` is different.

The full source code is available [here](https://github.com/Limeth/euclider/blob/6c9b2615796c6a558baf830336898e68ad10a3c1/src/util.rs#L657).

**Comment by kimsnj** [comment]
A similar false positive can be triggered within the `std` with `Cow<'a str>` and `Add` op.
[Link to playground](https://play.rust-lang.org/?gist=ff82365830f31e67269c55f62c7761f7&version=stable&mode=debug&edition=2015)

A check could be added for the commutative case `x = a + x` that `x` is a numerical type.
It would limit the usefulness of the lint for this branch but there would be no false positives.
 
Any thoughts?
